<a href="https://colab.research.google.com/github/nikitask14/cross-site-heart-disease-generalisation/blob/main/NN_Research_Capstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###**cross-site-heart-disease-generalisation**
**Stage1: Understanding the dataset**

In [1]:
import pandas as pd
import numpy as np
# Loaded the four raw site datasets — Cleveland, Hungary, Switzerland, and VA Long Beach — from UCI using pd.read_csv().
cleveland_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"
cleveland = pd.read_csv(cleveland_url, header = None)

hungarian_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.hungarian.data"
hungary = pd.read_csv(hungarian_url, header = None)

switzerland_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.switzerland.data"
switzerland = pd.read_csv(switzerland_url, header = None)

va_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.va.data"
va = pd.read_csv(va_url, header = None)


header=None

pandas had no column names(header) available, so it automatically labelled the 14 columns: 0, 1, 2, 3, ... 13

In [2]:
cleveland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       303 non-null    float64
 1   1       303 non-null    float64
 2   2       303 non-null    float64
 3   3       303 non-null    float64
 4   4       303 non-null    float64
 5   5       303 non-null    float64
 6   6       303 non-null    float64
 7   7       303 non-null    float64
 8   8       303 non-null    float64
 9   9       303 non-null    float64
 10  10      303 non-null    float64
 11  11      303 non-null    object 
 12  12      303 non-null    object 
 13  13      303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


In [3]:
hungary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       294 non-null    int64  
 1   1       294 non-null    int64  
 2   2       294 non-null    int64  
 3   3       294 non-null    object 
 4   4       294 non-null    object 
 5   5       294 non-null    object 
 6   6       294 non-null    object 
 7   7       294 non-null    object 
 8   8       294 non-null    object 
 9   9       294 non-null    float64
 10  10      294 non-null    object 
 11  11      294 non-null    object 
 12  12      294 non-null    object 
 13  13      294 non-null    int64  
dtypes: float64(1), int64(4), object(9)
memory usage: 32.3+ KB


In [4]:
switzerland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       123 non-null    int64 
 1   1       123 non-null    int64 
 2   2       123 non-null    int64 
 3   3       123 non-null    object
 4   4       123 non-null    int64 
 5   5       123 non-null    object
 6   6       123 non-null    object
 7   7       123 non-null    object
 8   8       123 non-null    object
 9   9       123 non-null    object
 10  10      123 non-null    object
 11  11      123 non-null    object
 12  12      123 non-null    object
 13  13      123 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 13.6+ KB


In [5]:
va.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       200 non-null    int64 
 1   1       200 non-null    int64 
 2   2       200 non-null    int64 
 3   3       200 non-null    object
 4   4       200 non-null    object
 5   5       200 non-null    object
 6   6       200 non-null    int64 
 7   7       200 non-null    object
 8   8       200 non-null    object
 9   9       200 non-null    object
 10  10      200 non-null    object
 11  11      200 non-null    object
 12  12      200 non-null    object
 13  13      200 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 22.0+ KB


**info()** only tells us pandas did not recognize any entries as NaN. But this raw dataset may use a symbol such as **?** to represent missing data.

Pandas would treat **?** as an ordinary string, which could explain why many columns in the 4 datasets became object.

In [6]:
cleveland.head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


**Note:**
1. The columns are just numbered 0 through 13. That is because the raw file contained no column names.
2. Before we inspect distributions, missing values, target, or anything else, we need to attach the correct variable names and understand them.

In [7]:
# Used the official data dictionary to map columns 0–13 to meaningful variable names rather than guessing from values.
column_names = ["age","sex", "cp", "trestbps", "chol", "fbs","restecg","thalach", "exang", "oldpeak", "slope", "ca", "thal","num"]
cleveland.columns = column_names
hungary.columns = column_names
switzerland.columns = column_names
va.columns = column_names



In [8]:
cleveland.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [9]:
hungary.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,28,1,2,130,132,0,2,185,0,0.0,?,?,?,0
1,29,1,2,120,243,0,0,160,0,0.0,?,?,?,0
2,29,1,2,140,?,0,0,170,0,0.0,?,?,?,0
3,30,0,1,170,237,0,1,170,0,0.0,?,?,6,0
4,31,0,2,100,219,0,1,150,0,0.0,?,?,?,0


In [10]:
switzerland.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,32,1,1,95,0,?,0,127,0,.7,1,?,?,1
1,34,1,4,115,0,?,?,154,0,.2,1,?,?,1
2,35,1,4,?,0,?,0,130,1,?,?,?,7,3
3,36,1,4,110,0,?,0,125,1,1,2,?,6,1
4,38,0,4,105,0,?,0,166,0,2.8,1,?,?,2


In [11]:
va.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63,1,4,140,260,0,1,112,1,3,2,?,?,2
1,44,1,4,130,209,0,1,127,0,0,?,?,?,0
2,60,1,4,132,218,0,1,140,1,1.5,3,?,?,2
3,55,1,4,142,228,0,1,149,1,2.5,1,?,?,1
4,66,1,3,110,213,1,2,99,1,1.3,2,?,?,0


**Compare the size of each site**

This matters because unequal site sizes are one possible source of heterogeneity and will matter later when we think about pooled training.

In [12]:
print(cleveland.shape)
print(hungary.shape)
print(switzerland.shape)
print(va.shape)

(303, 14)
(294, 14)
(123, 14)
(200, 14)




```
Cleveland     303 rows
Hungary       294 rows
Switzerland   123 rows
VA Long Beach 200 rows
```



All four have the same 14-column structure, but the number of patient records differs by site.

That is our first concrete sign of heterogeneity:

same schema
≠
same dataset size

In [13]:
cleveland.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    object 
 12  thal      303 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


Both **ca** and **thal** are supposed to contain numeric codes, yet pandas has stored them as object.

That suggests:

something non-numeric may be present in those columns

In [14]:
cleveland["ca"].unique()

array(['0.0', '3.0', '2.0', '1.0', '?'], dtype=object)

1. The ? is being used as a missing-value marker in the raw file.

2. Because ? is text, pandas cannot treat the whole column as numeric, so it stores the column as object.

3. This also corrects our earlier observation: 303 non-null

4. It did not mean “no missing data.” It only meant pandas had not yet recognised ? as missing.

In [15]:
cleveland["thal"].unique()

array(['6.0', '3.0', '7.0', '?'], dtype=object)

In [16]:
(cleveland["ca"] == "?").sum()

np.int64(4)

In [17]:
(cleveland["thal"] == '?').sum()

np.int64(2)

In [18]:
# clean way to do it
(cleveland == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


**Part B: Inspect Datatype of Hungarian Dataset**

In [19]:
hungary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    int64  
 1   sex       294 non-null    int64  
 2   cp        294 non-null    int64  
 3   trestbps  294 non-null    object 
 4   chol      294 non-null    object 
 5   fbs       294 non-null    object 
 6   restecg   294 non-null    object 
 7   thalach   294 non-null    object 
 8   exang     294 non-null    object 
 9   oldpeak   294 non-null    float64
 10  slope     294 non-null    object 
 11  ca        294 non-null    object 
 12  thal      294 non-null    object 
 13  num       294 non-null    int64  
dtypes: float64(1), int64(4), object(9)
memory usage: 32.3+ KB


We know that Hungary dataset has numerical values but columns 3,4,5,6,7,8,10,11 and 12 have object datatype. We should inspect for missing values.

? is only one possible missing-value marker. A raw dataset could use things like:


```
?
NA
N/A
-
unknown
blank cell
999
-1

```



In [20]:
for column in hungary.columns:
  print(column, hungary[column].unique())

age [28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51
 52 53 54 55 56 57 58 59 60 61 62 63 65 66]
sex [1 0]
cp [2 1 3 4]
trestbps ['130' '120' '140' '170' '100' '105' '110' '125' '150' '98' '112' '145'
 '190' '160' '115' '142' '180' '132' '135' '?' '108' '124' '113' '122'
 '92' '118' '106' '200' '138' '136' '128' '155']
chol ['132' '243' '?' '237' '219' '198' '225' '254' '298' '161' '214' '220'
 '160' '167' '308' '264' '166' '340' '209' '260' '211' '173' '283' '194'
 '223' '315' '275' '297' '292' '182' '200' '204' '241' '339' '147' '273'
 '307' '289' '215' '281' '250' '184' '245' '291' '295' '269' '196' '268'
 '228' '358' '201' '249' '266' '186' '207' '218' '412' '224' '238' '230'
 '163' '240' '280' '257' '263' '276' '284' '195' '227' '253' '187' '202'
 '328' '168' '216' '129' '190' '188' '179' '210' '272' '180' '100' '259'
 '468' '274' '320' '221' '309' '312' '171' '208' '246' '305' '217' '365'
 '344' '394' '256' '326' '277' '270' '229' '85' '347' '251' '222' '2

This systematic inspection tells us several useful things.

For Hungary, the suspicious non-numeric marker we actually observe is consistently:

?

In [21]:
(hungary == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,1
chol,23
fbs,8
restecg,1
thalach,1
exang,1
oldpeak,0


1. slope → 190 missing out of 294
2. ca    → 291 missing out of 294
3. thal  → 266 missing out of 294

**Part C: Inspect Datatype of Switzerland Dataset**

In [22]:
switzerland.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   age       123 non-null    int64 
 1   sex       123 non-null    int64 
 2   cp        123 non-null    int64 
 3   trestbps  123 non-null    object
 4   chol      123 non-null    int64 
 5   fbs       123 non-null    object
 6   restecg   123 non-null    object
 7   thalach   123 non-null    object
 8   exang     123 non-null    object
 9   oldpeak   123 non-null    object
 10  slope     123 non-null    object
 11  ca        123 non-null    object
 12  thal      123 non-null    object
 13  num       123 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 13.6+ KB


In [23]:
for column in switzerland.columns:
  print(column, switzerland[column].unique())

age [32 34 35 36 38 40 41 42 43 45 46 47 48 50 51 52 53 54 55 56 57 58 59 60
 61 62 63 64 65 66 67 68 69 70 72 73 74]
sex [1 0]
cp [1 4 3 2]
trestbps ['95' '115' '?' '110' '105' '100' '135' '150' '125' '145' '140' '155'
 '160' '120' '130' '165' '80' '180' '170' '200' '185']
chol [0]
fbs ['?' '0' '1']
restecg ['0' '?' '1' '2']
thalach ['127' '154' '130' '125' '166' '156' '179' '128' '150' '120' '144' '176'
 '99' '122' '145' '140' '138' '133' '113' '118' '149' '124' '110' '139'
 '92' '104' '170' '163' '60' '126' '82' '95' '115' '135' '141' '155' '83'
 '97' '98' '100' '148' '103' '121' '131' '182' '105' '175' '94' '119'
 '143' '63' '70' '77' '117' '123' '134' '72' '78' '109' '86' '114' '93'
 '67' '90' '108' '136' '?' '157']
exang ['0' '1' '?']
oldpeak ['.7' '.2' '?' '1' '2.8' '0' '-1.1' '1.6' '-1.5' '1.5' '2' '.5' '-.1'
 '-2.6' '2.1' '-.7' '2.2' '3' '.1' '.3' '-2' '-1' '1.8' '1.4' '2.6' '.9'
 '2.4' '1.1' '.4' '2.5' '1.7' '-.8' '-.5' '-.9' '3.7' '1.3']
slope ['1' '?' '2' '3']
ca ['?' '1' '

In [24]:
(switzerland == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,2
chol,0
fbs,75
restecg,1
thalach,1
exang,1
oldpeak,6


1. 123 patients
2. explicit "?" missingness in several variables
3. ca missing for 118/123
4. thal missing for 52/123
5. fbs missing for 75/123
6. AND cholesterol appears unavailable for the entire site

**Part D: Inspect Datatype of VA Dataset**


In [25]:
va.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   age       200 non-null    int64 
 1   sex       200 non-null    int64 
 2   cp        200 non-null    int64 
 3   trestbps  200 non-null    object
 4   chol      200 non-null    object
 5   fbs       200 non-null    object
 6   restecg   200 non-null    int64 
 7   thalach   200 non-null    object
 8   exang     200 non-null    object
 9   oldpeak   200 non-null    object
 10  slope     200 non-null    object
 11  ca        200 non-null    object
 12  thal      200 non-null    object
 13  num       200 non-null    int64 
dtypes: int64(5), object(9)
memory usage: 22.0+ KB


In [26]:
for column in va.columns:
  print(column, va[column].unique())

age [63 44 60 55 66 65 56 59 62 57 46 58 64 74 52 69 51 54 77 61 40 41 42 53
 68 67 72 75 49 35 43 48 50 45 76 70 71 38 37]
sex [1 0]
cp [4 3 2 1]
trestbps ['140' '130' '132' '142' '110' '120' '150' '180' '160' '126' '?' '128'
 '170' '152' '116' '124' '0' '122' '144' '154' '125' '104' '136' '134'
 '138' '178' '146' '135' '158' '106' '112' '102' '96' '172' '155' '156'
 '118' '100' '190' '114' '127']
chol ['260' '209' '218' '228' '213' '0' '236' '267' '166' '220' '177' '186'
 '100' '171' '230' '281' '203' '277' '233' '240' '153' '224' '316' '311'
 '270' '217' '214' '252' '339' '216' '276' '458' '241' '384' '297' '248'
 '308' '208' '227' '210' '245' '225' '198' '195' '161' '258' '235' '305'
 '223' '282' '349' '?' '160' '312' '283' '142' '211' '306' '222' '202'
 '197' '204' '274' '192' '298' '272' '200' '261' '181' '221' '175' '219'
 '310' '232' '273' '182' '292' '289' '193' '170' '369' '173' '271' '244'
 '285' '243' '237' '165' '287' '256' '264' '226' '207' '284' '337' '254'
 '300' '333' 

In [27]:
(va == "?").sum()

,0
age,0
sex,0
cp,0
trestbps,56
chol,7
fbs,7
restecg,0
thalach,53
exang,53
oldpeak,56


## First-Pass Multi-Site Dataset Inspection

The four processed UCI Heart Disease site datasets were loaded separately:

* Cleveland
* Hungary
* Switzerland
* VA Long Beach

All four datasets use the same 14-variable schema, but they differ in both sample size and data completeness.

Initial inspection included:

* dataset shapes,
* column names and data types,
* unique-value inspection,
* identification of `?` as an explicit missing-value marker,
* column-wise missing-value counts for each site.

An important observation is that sharing the same schema does **not** mean the sites contain equally complete data. Hungary, Switzerland, and VA Long Beach have substantial missingness in several variables, particularly `ca`, `thal`, and `slope`.

The Switzerland cholesterol column also requires further investigation because all observed values are `0`, suggesting that missingness may not always be represented explicitly by `?`.

No preprocessing or missing-value handling decisions have been made yet. This stage is strictly concerned with understanding the raw site data before modifying it.


In [28]:
# Standardised confirmed missing markers to NaN but deliberately did not impute yet.
cleveland = cleveland.replace("?",np.nan)
hungary = hungary.replace("?",np.nan)
switzerland = switzerland.replace("?",np.nan)
va = va.replace("?", np.nan)

In [29]:
switzerland["chol"] = switzerland["chol"].replace(0, np.nan)
va["chol"] = va["chol"].replace(0,np.nan)

In [30]:
# Verifying that the cleaning of "?" and "0" worked correctly

print((cleveland == "?").sum())
print((hungary == "?").sum())
print((switzerland == "?").sum())
print((va == "?").sum())
print((switzerland["chol"] == 0).sum())
print((va["chol"] == 0).sum())


age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64
0
0


In [31]:
#checking how many columns have NaN now. These are the same "?"
#that we replaced with NaN
print(cleveland.isna().sum())
print(hungary.isna().sum())
print(switzerland.isna().sum())
print(va.isna().sum())

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64
age           0
sex           0
cp            0
trestbps      1
chol         23
fbs           8
restecg       1
thalach       1
exang         1
oldpeak       0
slope       190
ca          291
thal        266
num           0
dtype: int64
age           0
sex           0
cp            0
trestbps      2
chol        123
fbs          75
restecg       1
thalach       1
exang         1
oldpeak       6
slope        17
ca          118
thal         52
num           0
dtype: int64
age           0
sex           0
cp            0
trestbps     56
chol          7
fbs           7
restecg       0
thalach      53
exang        53
oldpeak      56
slope       102
ca          198
thal        166
num           0
dtype: int64


In [32]:
for df in [cleveland, hungary, switzerland, va]:
  df[column_names] = df[column_names].apply(pd.to_numeric)

Converted the original UCI feature columns to numeric dtype after replacing documented missing-value markers with `NaN`.  

This ensures numerical preprocessing such as median imputation and standardisation operates on true numeric values.

In [33]:
cleveland.dtypes
cleveland[column_names].dtypes

,0
age,float64
sex,float64
cp,float64
trestbps,float64
chol,float64
fbs,float64
restecg,float64
thalach,float64
exang,float64
oldpeak,float64


### Stage 2: Understand the target

- original num values
- site-wise class balance
- binary target justified
- binary target created and verified

In [34]:
cleveland["num"].unique()

array([0, 2, 1, 3, 4])

In [35]:
hungary["num"].unique()

array([0, 1])

In [36]:
switzerland["num"].unique()

array([1, 3, 2, 0, 4])

In [37]:
va["num"].unique()

array([2, 0, 1, 3, 4])

Hungary only contains 0 and 1, while the other three contain 0–4, the concern that it raises if we tried to treat num as a 5-class target across all four sites is that we have no examples at all for classes 2, 3, and 4 in the hungary dataset.

That creates a serious cross-site comparability problem. A model evaluated on Hungary could never be tested on those classes there, because they simply do not exist in that site's labels.

Binary presence/absence gives all four sites a common, interpretable target and keeps the focus on cross-site generalisation, which is what we want to study.

A 5-class version would introduce a second research problem—unequal/missing class representation across sites—before we have even learned the centralized multi-site workflow.

Before recoding, we want to know the class distribution at each site:

0      → disease absent

1–4    → disease present

Why this matters: two sites can have the same target definition but very different proportions of positive and negative cases. That would be another form of site heterogeneity and could later affect both training and how we interpret metrics.

In [38]:
(cleveland["num"] == 0).sum()

np.int64(164)

So, out of 303 examples we have 164 as zero so (303 - 164) are 1- 4

In [39]:
(hungary["num"] == 0).sum()

np.int64(188)

Out of 294, 188 are zero.

In [40]:
(switzerland["num"] == 0).sum()

np.int64(8)

Out of 123, 8 are zero

In [41]:
(va["num"] == 0).sum()

np.int64(51)

Out of 200, 51 are 0.

Cleveland: 164 absent, 139 present

Hungary: 188 absent, 106 present

Switzerland: 8 absent, 115 present

VA Long Beach: 51 absent, 149 present


---

```
The proportions are very different across sites:

Cleveland: ~54% absent / 46% present

Hungary: ~64% absent / 36% present
Switzerland: ~7% absent / 93% present
VA: ~26% absent / 75% present

That is a major finding already: target distribution differs substantially by site.

```



In [42]:
cleveland["binary_target"] = (cleveland["num"]!= 0).astype(int)

In [43]:
hungary["binary_target"] = (hungary["num"]!= 0).astype(int)

In [44]:
switzerland["binary_target"] = (switzerland["num"] != 0 ).astype(int)

In [45]:
va["binary_target"] = (va["num"]!= 0).astype(int)

## Target Definition

The original UCI Heart Disease target, `num`, contains values from `0` to `4`.

Across the four sites:

* Cleveland, Switzerland, and VA Long Beach contain values `0–4`.
* Hungary contains only `0` and `1`.

For this capstone, the target is converted to a binary variable:

* `0` → heart disease absent
* `1` → heart disease present, corresponding to original `num` values `1–4`

The original `num` column is retained for traceability, while `binary_target` is used for modelling.

## Prediction Problem

The capstone therefore uses a **binary classification** task:

> Given the patient features, predict whether heart disease is absent or present.

Binary classification was chosen because it provides a common target definition across all four sites and keeps the study focused on **cross-site generalisation**, rather than introducing the additional complication of uneven availability of the original `1–4` target levels across sites.


In [46]:
# Check if binary conversion for the target label worked
print(cleveland["binary_target"].unique())
print(switzerland["binary_target"].unique())
print(hungary["binary_target"].unique())
print(va["binary_target"].unique())

[0 1]
[1 0]
[0 1]
[1 0]


## Research Problem

The UCI Heart Disease collection contains data from four clinical sites: Cleveland, Hungary, Switzerland, and VA Long Beach.

These sites are heterogeneous. They differ in sample size, data completeness, missing-value patterns, and target distributions. Cleveland, Switzerland, and VA Long Beach contain the original target values `0–4`, whereas Hungary contains only `0` and `1`.

To provide a common prediction task across all four sites, the original target is converted into a binary classification problem:

* `0` → heart disease absent
* `1–4` → heart disease present

The central research problem is therefore:

> **Does a model trained on centrally pooled data from four heterogeneous clinical sites generalise consistently when evaluated separately on each site?**

The capstone investigates **cross-site generalisation under centralized learning**, with particular attention to whether aggregate pooled performance may hide differences in site-specific performance.


In [47]:
# Without an extra column, an individual row no longer tells us
# whether it came from Cleveland, Hungary, Switzerland, or VA.
cleveland["site"] ="cleveland"
hungary["site"] = "hungary"
switzerland["site"] = "switzerland"
va["site"] = "va"

#Simple check to see if now datasets have a metadata column
va.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num,binary_target,site
0,63,1,4,140.0,260.0,0.0,1,112.0,1.0,3.0,2.0,NaN,NaN,2,1,va
1,44,1,4,130.0,209.0,0.0,1,127.0,0.0,0.0,NaN,NaN,NaN,0,0,va
2,60,1,4,132.0,218.0,0.0,1,140.0,1.0,1.5,3.0,NaN,NaN,2,1,va
3,55,1,4,142.0,228.0,0.0,1,149.0,1.0,2.5,1.0,NaN,NaN,1,1,va
4,66,1,3,110.0,213.0,1.0,2,99.0,1.0,1.3,2.0,NaN,NaN,0,0,va


In [48]:
pooled_data = pd.concat([cleveland, hungary, switzerland, va],
                        ignore_index = True)

ignore_index = True

As each original DataFrame has its own row index starting from 0.
After stacking, we would otherwise have repeated index values from each site if we don't use ingore_index = True.

In [49]:
pooled_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            920 non-null    float64
 1   sex            920 non-null    float64
 2   cp             920 non-null    float64
 3   trestbps       861 non-null    float64
 4   chol           767 non-null    float64
 5   fbs            830 non-null    float64
 6   restecg        918 non-null    float64
 7   thalach        865 non-null    float64
 8   exang          865 non-null    float64
 9   oldpeak        858 non-null    float64
 10  slope          611 non-null    float64
 11  ca             309 non-null    float64
 12  thal           434 non-null    float64
 13  num            920 non-null    int64  
 14  binary_target  920 non-null    int64  
 15  site           920 non-null    object 
dtypes: float64(13), int64(2), object(1)
memory usage: 115.1+ KB




```
920 rows = 303 + 294 + 123 + 200, so all four sites are represented.
16 columns = original 14 + binary_target + site.
Missingness is still preserved rather than silently filled.
Several columns are still object.
That is because many numeric values originally came in as strings when ? was present.
Replacing ? with NaN does not automatically convert those columns back to numeric.
```



In [50]:
# verify the site column in the pooled dataset to make sure
# each site's rows survived pooling correctly
print((pooled_data["site"] == "cleveland").sum())
print((pooled_data["site"] == "hungary").sum())
print((pooled_data["site"] == "switzerland").sum())
print((pooled_data["site"] == "va").sum())

303
294
123
200


## Site-Wise Train, Validation, and Test Split

To support cross-site evaluation, each site's data is split separately before the corresponding training, validation, and test portions are pooled.

For each site:

* `60%` of examples are used for training.
* `20%` are used for validation.
* `20%` are reserved for final testing.

The split is performed in two stages:

1. Separate `20%` of the site data as the test set.
2. Split the remaining `80%` into training and validation data, using `25%` of that portion for validation.

This produces the final `60/20/20` split.

The binary target is stratified during both splitting stages so that the proportion of disease-absent and disease-present examples is preserved as closely as possible within each subset.

The predictor set contains the 13 clinical variables. The following columns are excluded from the model inputs:

* `num` — original target retained for traceability
* `binary_target` — the target to be predicted
* `site` — metadata retained for site-specific evaluation

Splitting each site separately ensures that Cleveland, Hungary, Switzerland, and VA Long Beach are represented in the training, validation, and test sets. The corresponding site-specific subsets can later be pooled for centralized model training while still allowing evaluation on each site separately.


In [51]:
from sklearn.model_selection import train_test_split

In [52]:
feature_columns = [
    "age", "sex", "cp", "trestbps", "chol", "fbs",
    "restecg", "thalach", "exang", "oldpeak",
    "slope", "ca", "thal"
]



In [53]:
X_cleveland = cleveland[feature_columns]
y_cleveland = cleveland["binary_target"]

x_train_val, x_test_cleveland, y_train_val, y_test_cleveland = train_test_split(
    X_cleveland,
    y_cleveland,
    test_size = 0.20,
    random_state = 42,
    stratify = y_cleveland

)

x_train_cleveland, x_val_cleveland, y_train_cleveland, y_val_cleveland = train_test_split(
    x_train_val,
    y_train_val,
    test_size = 0.25,
    random_state = 42,
    stratify = y_train_val
)

In [54]:
X_hungary = hungary[feature_columns]
y_hungary = hungary["binary_target"]

x_train_val, x_test_hungary, y_train_val, y_test_hungary = train_test_split(
    X_hungary,
    y_hungary,
    test_size = 0.20,
    random_state = 42,
    stratify = y_hungary

)

x_train_hungary, x_val_hungary, y_train_hungary, y_val_hungary = train_test_split(
    x_train_val,
    y_train_val,
    test_size = 0.25,
    random_state = 42,
    stratify = y_train_val
)

In [55]:
X_switzerland = switzerland[feature_columns]
y_switzerland = switzerland["binary_target"]

x_train_val, x_test_switzerland, y_train_val, y_test_switzerland = train_test_split(
    X_switzerland,
    y_switzerland,
    test_size = 0.20,
    random_state = 42,
    stratify = y_switzerland

)

x_train_switzerland, x_val_switzerland, y_train_switzerland, y_val_switzerland = train_test_split(
    x_train_val,
    y_train_val,
    test_size = 0.25,
    random_state = 42,
    stratify = y_train_val
)

In [56]:
X_va = va[feature_columns]
y_va = va["binary_target"]

x_train_val, x_test_va, y_train_val, y_test_va = train_test_split(
    X_va,
    y_va,
    test_size = 0.20,
    random_state = 42,
    stratify = y_va

)

x_train_va, x_val_va, y_train_va, y_val_va = train_test_split(
    x_train_val,
    y_train_val,
    test_size = 0.25,
    random_state = 42,
    stratify = y_train_val
)

In [57]:
# Verify shape of each of these
print(x_train_va.shape)
print(x_val_va.shape)
print(y_train_va.shape)
print(y_val_va.shape)
print(x_test_va.shape)
print(y_test_va.shape)


(120, 13)
(40, 13)
(120,)
(40,)
(40, 13)
(40,)


In [58]:
print(y_train_va.value_counts())
print("--")
print(y_val_va.value_counts())
print("--")
print(y_test_va.value_counts())

binary_target
1    89
0    31
Name: count, dtype: int64
--
binary_target
1    30
0    10
Name: count, dtype: int64
--
binary_target
1    30
0    10
Name: count, dtype: int64


## Verifying Split Sizes and Class Balance

After creating the site-specific train, validation, and test splits, two checks are performed.

### 1. Verify split sizes

The feature matrices (`X`) are checked using their shapes to confirm that the expected number of examples and predictor columns are present in each subset.

For VA Long Beach, the final split contains:

* 120 training examples
* 40 validation examples
* 40 test examples

with 13 predictor variables in each feature matrix.

### 2. Verify class balance

Class balance is checked using the target vectors (`y`), because they contain the binary labels:

* `0` → heart disease absent
* `1` → heart disease present

The feature matrices (`X`) are not used for this check because counting values in `X` would count repeated combinations of predictor values rather than the distribution of target classes.

For VA Long Beach:

* Training: 89 positive, 31 negative
* Validation: 30 positive, 10 negative
* Test: 30 positive, 10 negative

This confirms that stratification preserved the original class distribution approximately across the train, validation, and test subsets.


In [59]:
# Cleveland
print("CLEVELAND")
print("Shapes:")
print(x_train_cleveland.shape, y_train_cleveland.shape)
print(x_val_cleveland.shape, y_val_cleveland.shape)
print(x_test_cleveland.shape, y_test_cleveland.shape)

print("\nClass balance:")
print(y_train_cleveland.value_counts())
print("--")
print(y_val_cleveland.value_counts())
print("--")
print(y_test_cleveland.value_counts())


# Hungary
print("\nHUNGARY")
print("Shapes:")
print(x_train_hungary.shape, y_train_hungary.shape)
print(x_val_hungary.shape, y_val_hungary.shape)
print(x_test_hungary.shape, y_test_hungary.shape)

print("\nClass balance:")
print(y_train_hungary.value_counts())
print("--")
print(y_val_hungary.value_counts())
print("--")
print(y_test_hungary.value_counts())


# Switzerland
print("\nSWITZERLAND")
print("Shapes:")
print(x_train_switzerland.shape, y_train_switzerland.shape)
print(x_val_switzerland.shape, y_val_switzerland.shape)
print(x_test_switzerland.shape, y_test_switzerland.shape)

print("\nClass balance:")
print(y_train_switzerland.value_counts())
print("--")
print(y_val_switzerland.value_counts())
print("--")
print(y_test_switzerland.value_counts())

CLEVELAND
Shapes:
(181, 13) (181,)
(61, 13) (61,)
(61, 13) (61,)

Class balance:
binary_target
0    98
1    83
Name: count, dtype: int64
--
binary_target
0    33
1    28
Name: count, dtype: int64
--
binary_target
0    33
1    28
Name: count, dtype: int64

HUNGARY
Shapes:
(176, 13) (176,)
(59, 13) (59,)
(59, 13) (59,)

Class balance:
binary_target
0    112
1     64
Name: count, dtype: int64
--
binary_target
0    38
1    21
Name: count, dtype: int64
--
binary_target
0    38
1    21
Name: count, dtype: int64

SWITZERLAND
Shapes:
(73, 13) (73,)
(25, 13) (25,)
(25, 13) (25,)

Class balance:
binary_target
1    69
0     4
Name: count, dtype: int64
--
binary_target
1    23
0     2
Name: count, dtype: int64
--
binary_target
1    23
0     2
Name: count, dtype: int64


The main thing that stands out is Switzerland:

train: 4 negatives, 69 positives

val:   2 negatives, 23 positives

test:  2 negatives, 23 positives

So the split is technically correct, but the site itself is extremely imbalanced. That means later, a Switzerland metric based on only 2 negative test examples will be unstable and must be interpreted cautiously.

###**Pool Data training, validation and test wise**

In [60]:
pooled_train_x = pd.concat([x_train_cleveland, x_train_hungary, x_train_switzerland,x_train_va],ignore_index = True)
pooled_train_y = pd.concat([y_train_cleveland, y_train_hungary, y_train_switzerland, y_train_va],ignore_index = True)
pooled_val_x = pd.concat([x_val_cleveland,x_val_hungary, x_val_switzerland, x_val_va],ignore_index = True)
pooled_val_y = pd.concat([y_val_cleveland,y_val_hungary,y_val_switzerland, y_val_va],ignore_index = True)
pooled_test_x = pd.concat([x_test_cleveland,x_test_hungary, x_test_switzerland,x_test_va],ignore_index = True)
pooled_test_y = pd.concat([y_test_cleveland, y_test_hungary,y_test_switzerland, y_test_va],ignore_index = True)

verify the shapes of the pooled train, validation, and test sets

In [61]:
print(pooled_train_x.shape)
print(pooled_train_y.shape)
print(pooled_val_x.shape)
print(pooled_val_y.shape)
print(pooled_test_x.shape)
print(pooled_test_y.shape)

(550, 13)
(550,)
(185, 13)
(185,)
(185, 13)
(185,)


##**Stage 2: Preprocessing**

In [62]:
numerical_columns_ = ["age","trestbps", "chol","thalach","oldpeak", "ca"]
categorical_columns_ = ["sex", "cp", "fbs", "restecg", "exang","slope", "thal"]

So relative to published practice, our thinking is quite reasonable:

ca → strong precedent to exclude

thal → strong precedent to exclude

slope → also commonly excluded when combining sites

chol → more debatable

In [63]:
numerical_columns = ["age","trestbps","thalach","oldpeak"]
categorical_columns = ["sex", "cp", "restecg", "exang"]

In [64]:
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import (OneHotEncoder,StandardScaler)

In [65]:
numerical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy = "median")),
    ("scalar", StandardScaler())
])

In [66]:
categorical_pipeline = Pipeline([
   ("imputer", SimpleImputer(strategy = "most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown = "ignore"))

])

In [67]:
preprocessor = ColumnTransformer([
    ("numerical" , numerical_pipeline, numerical_columns),
    ("categorical", categorical_pipeline, categorical_columns)
])

##**Stage 3: Lock the Basline Model - Logistic Regression**

In [68]:
from sklearn.linear_model import LogisticRegression

In [69]:
baseline_model = Pipeline([
    ("preprocessor", preprocessor),
    ("classification", LogisticRegression(max_iter = 1000) )
])

baseline_model

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('numerical',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scalar',
                                                                   StandardScaler())]),
                                                  ['age', 'trestbps', 'thalach',
                                                   'oldpeak']),
                                                 ('categorical',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['sex', 'cp', 'restecg',
                                                   'exang'])])),
                ('classification', LogisticRegression(max_iter=1000))])

In [70]:
baseline_model.fit(pooled_train_x, pooled_train_y)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('numerical',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scalar',
                                                                   StandardScaler())]),
                                                  ['age', 'trestbps', 'thalach',
                                                   'oldpeak']),
                                                 ('categorical',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['sex', 'cp', 'restecg',
                                                   'exang'])])),
                ('classification', LogisticRegression(max_iter=1000))])

###**Pooled Validation**

In [71]:
prediction_val = baseline_model.predict(pooled_val_x)
prediction_val.shape

(185,)

In [72]:
from sklearn.metrics import (confusion_matrix, precision_score, recall_score, accuracy_score, f1_score)

In [73]:
baseline_cm = confusion_matrix(pooled_val_y, prediction_val)
baseline_cm

array([[67, 16],
       [20, 82]])

In [74]:
baseline_accuracy = accuracy_score(pooled_val_y, prediction_val)
baseline_accuracy

0.8054054054054054

In [75]:
baseline_precision = precision_score(pooled_val_y, prediction_val)
baseline_precision

0.8367346938775511

In [76]:
baseline_recall = recall_score(pooled_val_y, prediction_val)
baseline_recall

0.803921568627451

In [77]:
baseline_f1 = f1_score(pooled_val_y, prediction_val)
baseline_f1

0.82

###**Site-wise Validation**

In [78]:
# Manual Cleveland validation
prediction_cleveland = baseline_model.predict(x_val_cleveland)

baseline_cm_cleveland = confusion_matrix(y_val_cleveland, prediction_cleveland )
print(baseline_cm_cleveland)

baseline_accuracy_cleveland = accuracy_score(y_val_cleveland, prediction_cleveland )
print(baseline_accuracy_cleveland)

baseline_precision_cleveland = precision_score(y_val_cleveland, prediction_cleveland )
print(baseline_precision_cleveland)

baseline_recall_cleveland = recall_score(y_val_cleveland, prediction_cleveland )
print(baseline_recall_cleveland)

baseline_f1_cleveland = f1_score(y_val_cleveland, prediction_cleveland )
print(baseline_f1_cleveland)

[[29  4]
 [ 5 23]]
0.8524590163934426
0.8518518518518519
0.8214285714285714
0.8363636363636363


In [79]:
def evaluate_site(model, x, y):
  prediction = model.predict(x)
  cm = confusion_matrix(y, prediction)
  accuracy = accuracy_score(y, prediction)
  precision = precision_score(y, prediction)
  recall = recall_score(y, prediction)
  f1score = f1_score(y, prediction)


  print("Confusion Matrix:", cm)
  print("Accuracy:", accuracy)
  print("Precision", precision)
  print("Recall:", recall)
  print("F1 Score:", f1score)

evaluate_site(baseline_model, x_val_hungary,y_val_hungary )
print("")
print("")
evaluate_site(baseline_model, x_val_switzerland,y_val_switzerland )
print("")
print("")
evaluate_site(baseline_model, x_val_va,y_val_va )


Confusion Matrix: [[32  6]
 [ 7 14]]
Accuracy: 0.7796610169491526
Precision 0.7
Recall: 0.6666666666666666
F1 Score: 0.6829268292682927


Confusion Matrix: [[ 2  0]
 [ 2 21]]
Accuracy: 0.92
Precision 1.0
Recall: 0.9130434782608695
F1 Score: 0.9545454545454546


Confusion Matrix: [[ 4  6]
 [ 6 24]]
Accuracy: 0.7
Precision 0.8
Recall: 0.8
F1 Score: 0.8


**Observations:**

1. On the Cleveland validation set, the pooled model shows relatively balanced performance across the two classes, with similar numbers of false positives and false negatives and an F1 score of 0.836.

2. Hungary shows weaker validation performance than Cleveland. The main concern is lower sensitivity to the positive class: 7 of 21 disease-present cases are missed, giving a recall of 0.667.

3. Switzerland has an extremely imbalanced validation set (2 negative, 23 positive cases). Although accuracy (0.92), precision (1.00), recall (0.913), and F1 (0.955) appear high, these values must be interpreted cautiously. A model predicting every patient as positive would already achieve 92% accuracy. The model correctly classified both available negative cases, but two examples are insufficient to establish reliable performance on the negative class.

4. VA has a moderately imbalanced validation set (10 negative, 30 positive). The model detects 24 of 30 positive cases, giving recall of 0.80, but performs poorly on the negative class, correctly identifying only 4 of 10 negatives. This contributes to the relatively low overall accuracy of 0.70.

### Overall validation interpretation

The pooled logistic-regression model does not perform consistently across the four sites. Hungary shows weaker positive-class detection, with a recall of 0.667, while VA shows poor recognition of the negative class, correctly identifying only 4 of 10 negative cases.

Switzerland reports very high validation metrics, but these must be interpreted cautiously because its validation set is extremely imbalanced (2 negative vs 23 positive cases). A model predicting every patient as positive would already achieve 92% accuracy, so accuracy alone is not informative for this site.

Overall, the reasonably strong pooled validation metrics do not imply consistent performance across individual sites. Site-wise evaluation reveals differences that would be hidden if only pooled performance were reported.

**Test**

In [88]:
test_predictions = baseline_model.predict(pooled_test_x)

In [89]:
test_cm = confusion_matrix(pooled_test_y,test_predictions)
print(test_cm)

test_accuracy = accuracy_score(pooled_test_y,test_predictions)
print(test_accuracy)

test_precision = precision_score(pooled_test_y,test_predictions)
print(test_precision)

test_recall = recall_score(pooled_test_y,test_predictions)
print(test_recall)

test_f1 = f1_score(pooled_test_y,test_predictions)
print(test_f1)

[[66 17]
 [10 92]]
0.8540540540540541
0.8440366972477065
0.9019607843137255
0.8720379146919431


Now we compare the pooled validation results with the pooled test results


```
| Metric    | Validation |    Test   |
| --------- | ---------  | --------- |
| Accuracy  |      0.805 |  0.854    |
| Precision |      0.837 |  0.844    |
| Recall    |      0.804 |  0.902    |
| F1        |      0.820 |  0.872    |

```



In [91]:
def site_wise_test_evaluation(model, x, y):
  predictions = model.predict(x)
  cm = confusion_matrix(y, predictions)
  accuracy = accuracy_score(y, predictions)
  precision = precision_score(y, predictions)
  recall = recall_score(y, predictions)
  f1 = f1_score(y, predictions)

  print("Confusion Matrix:", cm)
  print("Accuracy:", accuracy)
  print("Precision:", precision)
  print("Recall:", recall)
  print("F1:", f1)


site_wise_test_evaluation(baseline_model,x_test_cleveland, y_test_cleveland)
site_wise_test_evaluation(baseline_model,x_test_hungary, y_test_hungary)
site_wise_test_evaluation(baseline_model,x_test_switzerland, y_test_switzerland)
site_wise_test_evaluation(baseline_model,x_test_va, y_test_va)


Confusion Matrix: [[28  5]
 [ 2 26]]
Accuracy: 0.8852459016393442
Precision: 0.8387096774193549
Recall: 0.9285714285714286
F1: 0.8813559322033898
Confusion Matrix: [[33  5]
 [ 3 18]]
Accuracy: 0.864406779661017
Precision: 0.782608695652174
Recall: 0.8571428571428571
F1: 0.8181818181818182
Confusion Matrix: [[ 0  2]
 [ 4 19]]
Accuracy: 0.76
Precision: 0.9047619047619048
Recall: 0.8260869565217391
F1: 0.8636363636363636
Confusion Matrix: [[ 5  5]
 [ 1 29]]
Accuracy: 0.85
Precision: 0.8529411764705882
Recall: 0.9666666666666667
F1: 0.90625


### Validation → Test Performance by Site

| Site        | Accuracy        | Precision       | Recall          | F1              |
|-------------|-----------------|-----------------|-----------------|-----------------|
| Cleveland   | .852 → **.885** | .852 → .839     | .821 → **.929** | .836 → **.881** |
| Hungary     | .780 → **.864** | .700 → **.783** | .667 → **.857** | .683 → **.818** |
| Switzerland | .920 → **.760** | 1.000 → .905    | .913 → .826     | .955 → .864     |
| VA          | .700 → **.850** | .800 → .853     | .800 → **.967** | .800 → .906     |

*Values show validation → test performance for the same centrally trained logistic-regression baseline.*

### Final cross-site test interpretation

1. The pooled logistic-regression baseline shows reasonably strong overall performance on the untouched pooled test set (accuracy = 0.854, precision = 0.844, recall = 0.902, F1 = 0.872), with no substantial validation-to-test performance drop.

2. However, the site-wise results show that this pooled performance does not translate into completely consistent behaviour across all sites. Cleveland shows relatively stable and strong performance, while Hungary improves substantially on the test split, particularly in positive-class recall. VA also improves markedly, with only one false negative and a recall of 0.967 on the test set.

3. Switzerland remains the most uncertain site. Its test performance is substantially lower than its validation performance, and both splits contain only 2 negative cases compared with 23 positive cases. The model correctly classified both negatives during validation but neither during testing, demonstrating how unstable site-level conclusions can be when one class is represented by very few examples.

4. Overall, the centrally pooled model generalises reasonably well to unseen patients from several sites, particularly for detecting the positive class. However, performance varies across sites and the reliability of site-specific estimates depends strongly on local sample size and class distribution. Therefore, strong pooled performance alone is not sufficient evidence of consistent cross-site generalisation.